types/react-dom/index.d.ts

```diff
@@ -29,7 +29,7 @@ import {
 export function findDOMNode(instance: ReactInstance | null | undefined): Element | null | Text;
 export function unmountComponentAtNode(container: Element | DocumentFragment): boolean;
 
-export function createPortal(children: ReactNode, container: Element, key?: null | string): ReactPortal;
+export function createPortal(children: ReactNode, container: Element | ShadowRoot, key?: null | string): ReactPortal;
```

Comment:
The docs you linked specifically say it's a DOM element:

> The second argument (container) is a DOM element.
-- https://reactjs.org/docs/portals.html

Could you clarify that this is officially supported by React? Just because code is seemingly working, doesn't mean it's supported.

Reply:
Thanks for the comment — having looked further into it, it seems actually `Element | ShadowRoot` is still too restrictive.

[`createPortal`](https://github.com/facebook/react/blob/cae635054e17a6f107a39d328649137b83f25972/packages/react-dom/src/client/ReactDOM.js#L108-L120) validates using [`isValidContainer`](https://github.com/facebook/react/blob/cae635054e17a6f107a39d328649137b83f25972/packages/react-dom/src/client/ReactDOMRoot.js#L224-L231), which checks if the node is a `Container` (`nodeType` is `ELEMENT_NODE`, `DOCUMENT_NODE`, or `DOCUMENT_FRAGMENT_NODE`). ShadowRoot inherits from `DocumentFragment`, so its `nodeType` is `DOCUMENT_FRAGMENT_NODE`.

I've updated the PR accordingly. I notice there are various other functions in React DOM that take a `container` argument with a type narrower than `Container`, but I've left those out of the scope of this PR for now.

Reply:
> [createPortal](https://github.com/facebook/react/blob/cae635054e17a6f107a39d328649137b83f25972/packages/react-dom/src/client/ReactDOM.js#L108-L120) validates using [isValidContainer](https://github.com/facebook/react/blob/cae635054e17a6f107a39d328649137b83f25972/packages/react-dom/src/client/ReactDOMRoot.js#L224-L231), which checks if the node is a Container (nodeType is ELEMENT_NODE, DOCUMENT_NODE, or DOCUMENT_FRAGMENT_NODE). ShadowRoot inherits from DocumentFragment, so its nodeType is DOCUMENT_FRAGMENT_NODE.

I would wait for https://github.com/facebook/react/pull/21728 before moving along. Code seems to support your assumptions but just because the implementation currently supports something does not imply its intended usage.

Reply:
@eps1lon OK sure, makes sense 👍